src/weights.jl

```diff
@@ -2,44 +2,189 @@
+"""
+macro weights(name)
+    return quote
+        if VERSION < v"0.6.0-dev.2123"
+            immutable $name{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        else
+            immutable $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        end
     end
 end
 
+eltype(wv::AbstractWeights) = eltype(wv.values)
+length(wv::AbstractWeights) = length(wv.values)
+values(wv::AbstractWeights) = wv.values
+sum(wv::AbstractWeights) = wv.sum
+isempty(wv::AbstractWeights) = isempty(wv.values)
+
+Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
+Base.size(wv::AbstractWeights) = size(wv.values)
+
+"""
+    cfactor(n::Integer, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
+If `corrected=true` this will return ``\\frac{1}{n - 1}``
+(ie: [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise it will return ``\\frac{1}{n}``.
+"""
+cfactor(n::Integer, corrected=false) = 1 / (n - Int(corrected))
+
+"""
+    cfactor(wv::AbstractWeights, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with a set of
+weights `wv`.
+"""
+cfactor(wv::AbstractWeights, corrected=false) = cfactor(wv, Val{corrected})
+
+"""
+    cfactor(wv::AbstractWeights, false)
+
+``\\frac{1}{\sum w}``
+"""
+cfactor(wv::AbstractWeights, ::Type{Val{false}}) = 1 / sum(wv)
+cfactor(wv::AbstractWeights, ::Type{Val{true}}) =
+    throw(ArgumentError("$(typeof(wv)) does not support bias correction."))
+
+
+@weights AnalyticWeights
+
+"""
+    AnalyticWeights(vs, wsum=sum(vs))
+
+Construct an `AnalyticWeights` vector with weight values `vs` and sum of weights `wsum`.
+
+Analytic weights describe a non-random relative importance (usually between 0 and 1)
+for each observation. These weights may also be referred to as reliability weights,
+precision weights or inverse variance weights.
+"""
+AnalyticWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) =
+    AnalyticWeights{S, eltype(vs), V}(vs, s)
+
+"""
+    aweights(vs)
+
+Construct an `AnalyticWeights` vector from a given array.
+See the documentation for `AnalyticWeights` for more details.
 """
-    WeightVec(vs, [wsum])
+aweights(vs::RealVector) = AnalyticWeights(vs)
+aweights(vs::RealArray) = AnalyticWeights(vec(vs))
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
-If omitted, `wsum` is computed.
 """
-function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
-    return WeightVec{S, eltype(vs), V}(vs, s)
+    cfactor(w::AnalyticWeights, true)
+
+``\\frac{1}{\sum w - \sum w / \sum {w^2}}``
+"""
+function cfactor(w::AnalyticWeights, ::Type{Val{true}})
+    s = sum(w)
+    sum_sn = 0.0
+    for x in w
+        sum_sn += (x / s) ^ 2
+    end
+
+    1 / (s * (1 - sum_sn))
```

Comment:
The accidentally wrote the division backwards in the formula. According to the wiki page it should be `1 / (V1 - (V2 / V1^2)` where `V1` is the sum of the weights and `V2` is the sum of the squared weights. If the weights are normalized to sum to 1 then we can simplify it to `1 / (1 - V2)`. The we just scale the denominator by the actual scale of the weights, so that we don't need the weights to be normalized.